ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+
+        completions = await func(*args, **kwargs)
+        if isinstance(instance, langchain.llms.OpenAI):
+            _tag_openai_token_usage(span, completions.llm_output)
+            integration.record_usage(span, completions.llm_output)
+
+        for idx, completion in enumerate(completions.generations):
+            if integration.is_pc_sampled_span(span):
+                span.set_tag_str("langchain.response.completions.%d.text" % idx, integration.trunc(completion[0].text))
+            if completion[0].generation_info is not None:
+                span.set_tag_str(
+                    "langchain.response.completions.%d.finish_reason" % idx,
+                    str(completion[0].generation_info.get("finish_reason")),
+                )
+                span.set_tag_str(
+                    "langchain.response.completions.%d.logprobs" % idx,
+                    str(completion[0].generation_info.get("logprobs")),
+                )
+        if integration.is_pc_sampled_log(span):
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s" % (instance.__module__, instance.__class__.__name__),
+                attrs={
+                    "prompts": prompts,
+                    "choices": [
+                        [{"text": completion.text} for completion in completions]
+                        for completions in completions.generations
+                    ],
+                },
+            )
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        integration.metric(span, "incr", "request.error", 1)
+        raise
+    finally:
+        span.finish()
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
+    return completions
+
+
+@with_traced_module
+def traced_chat_model_generate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type.split("-")[0]
+    chat_messages = args[0]
+    integration = langchain._datadog_integration
+    span = integration.trace(pin, "%s.%s" % (instance.__module__, instance.__class__.__name__))
+    try:
+        for message_set_idx, message_set in enumerate(chat_messages):
+            for message_idx, message in enumerate(message_set):
+                if integration.is_pc_sampled_span(span):
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.%d.content" % (message_set_idx, message_idx),
+                        integration.trunc(message.content),
+                    )
+                span.set_tag_str(
+                    "langchain.request.messages.%d.%d.message_type" % (message_set_idx, message_idx),
+                    message.__class__.__name__,
+                )
+        model = _extract_model_name(instance)
+        if model is not None:
+            span.set_tag_str("langchain.request.model", model)
+        span.set_tag_str("langchain.request.provider", llm_provider)
```

Comment:
This seems to be a pattern that's repeated a lot. You might consider deduplicating it to make changes easier.

Reply:
Good point. I've added a helper `_set_base_span_tags()` to set these base level tags for all spans.